paprika/tleap.py

```diff
@@ -0,0 +1,574 @@
+            if self.pbc_type == 'cubic':
+                f.write("solvatebox {} {} {} iso\n".format(self.unit, self.water_box, self.buffer_value))
+            elif self.pbc_type == 'rectangular':
+                f.write("solvatebox {} {} {{10.0 10.0 {}}}\n".format(self.unit, self.water_box, self.buffer_value))
+            elif self.pbc_type == 'octahedral':
+                f.write("solvateoct {} {} {} iso\n".format(self.unit, self.water_box, self.buffer_value))
+            elif self.pbc_type is None:
+                f.write("# Skipping solvation ...\n")
+            else:
+                raise Exception(
+                    "Incorrect pbctype value provided: " + str(self.pbc_type) \
+                    + ". Only `cubic`, `rectangular`, `octahedral`, and None are valid")
+            if self.neutralize:
+                f.write("addionsrand {} {} 0\n".format(self.unit, self.counter_cation))
+                f.write("addionsrand {} {} 0\n".format(self.unit, self.counter_anion))
+            # Additional ions should be specified as a list, with residue name and number of ions in pairs, like ['NA',
+            # 5] for five additional sodium ions. By this point, if the user specified a molality or molarity,
+            # it should already have been converted into a number.
+            if self.add_ion_residues:
+                for residue, amount in zip(self.add_ion_residues[0::2],
+                                           self.add_ion_residues[1::2]):
+                    f.write("addionsrand {} {} {}\n".format(self.unit, residue, amount))
+            if self.waters_to_remove:
+                for water_number in self.waters_to_remove:
+                    f.write("remove {} {}.{}\n".format(self.unit, self.unit, water_number))
+    
+            # Note, the execution of tleap is assumed to take place in the
+            # same directory as all the associated input files, so we won't
+            # put directory paths on the saveamberparm or savepdb commands.
+            if self.output_prefix and include_saves:
+                f.write("savepdb {} {}.pdb\n".format(self.unit, self.output_prefix))
+                f.write("saveamberparm {} {}.prmtop {}.rst7\n".format(self.unit, self.output_prefix,
+                                                                      self.output_prefix))
+            else:
+                pass
+            f.write("desc {}\n".format(self.unit))
+            f.write("quit\n")
+
+    def run(self):
+        """
+        Execute `tleap`
+    
+        """
+    
+        self.check_for_leap_log()
+        file_name = self.output_prefix + '.tleap.in'    
+        p = sp.Popen(['tleap', '-s ', '-f ', file_name], stdout=sp.PIPE, bufsize=1, universal_newlines=True, cwd=self.output_path)
+        output = []
+        # Wait until process terminates...
+        while p.poll() is None:
+            line = p.communicate()[0]
+            output.append(line)
+        if p.poll() is None:
+            p.kill()
```

Comment:
I half wonder if we should just put a sleep command here and kill `tleap` if it's been running for more than a few seconds.

Reply:
Which one will execute faster?

Reply:
I think I meant using `sp.call` or one or the others, instead of the `communicate()` method which will just let it run forever.

Reply:
Not sure what exactly you are proposing ... leaving alone for now as I haven't seen any issues with it.